off/off.go

```diff
@@ -0,0 +1,202 @@
+	writer.FileFormat = "OFF"
+	writer.FileFormatVersion = "0.1.0"
+	writer.MaxPresamples = MaxPresamples
+	writer.FramePeriodSeconds = FramePeriodSeconds
+	writer.NumberOfBases, _ = Projectors.Dims()
+	writer.ModelInfo = ModelInfo{Projectors: *NewArrayJsoner(Projectors), Basis: *NewArrayJsoner(Basis),
+		Description: ModelDescription}
+	writer.CreationInfo = CreationInfo{DastardVersion: DastardVersion, GitHash: GitHash,
+		SourceName: SourceName, CreationTime: time.Now()}
+	writer.ReadoutInfo = ReadoutInfo
+	writer.fileName = fileName
+	return writer
+}
+
+// ModelInfo stores info related to the model (aka basis, aka projectors) for printing to the file header, aids with json formatting
+type ModelInfo struct {
+	Projectors  ArrayJsoner
+	Basis       ArrayJsoner
+	Description string
+}
+
+// ArrayJsoner aids in formatting arrays for writing to JSON
+type ArrayJsoner struct {
+	RowMajorFloat64ValuesBase64 string
+	Rows                        int
+	Cols                        int
+}
+
+// NewArrayJsoner creates an ArrayJsoner from a mat.Dense
+func NewArrayJsoner(array *mat.Dense) *ArrayJsoner {
+	v := new(ArrayJsoner)
+	v.Rows, v.Cols = array.Dims()
+	v.RowMajorFloat64ValuesBase64 = base64.StdEncoding.EncodeToString(getbytes.FromSliceFloat64(array.RawMatrix().Data))
+	return v
+}
+
+// CreationInfo stores info related to file creation for printing to the file header, aids with json formatting
+type CreationInfo struct {
+	DastardVersion string
+	GitHash        string
+	SourceName     string
+	CreationTime   time.Time
+}
+
+// TimeDivisionMultiplexingInfo stores info related to tdm readout for printing to the file header, aids with json formatting
+type TimeDivisionMultiplexingInfo struct {
+	NumberOfRows    int
+	NumberOfColumns int
+	NumberOfChans   int
+	ColumnNum       int
+	RowNum          int
+}
+
+// HeaderWritten returns true if header has been written.
+func (w *Writer) HeaderWritten() bool {
+	return w.headerWritten
+}
+
+// RecordsWritten return the nunber of records written.
+func (w *Writer) RecordsWritten() int {
+	return w.recordsWritten
+}
+
+// WriteHeader writes a header to the file
+func (w *Writer) WriteHeader() error {
+	if w.headerWritten {
+		return errors.New("header already written")
+	}
+	s, err := json.MarshalIndent(w, "", "    ")
+	if err != nil {
+		return err
+	}
+	if _, err1 := w.writer.Write(s); err != nil {
+		return err1
+	}
+	if _, err1 := w.writer.WriteString("\n"); err != nil {
+		return err1
+	}
+	w.headerWritten = true
+	return nil
+}
+
+// WriteRecord writes a record to the file
+func (w *Writer) WriteRecord(recordSamples int32, recordPreSamples int32, framecount int64,
+	timestamp int64, pretriggerMean float32, residualStdDev float32, data []float32) error {
+	if len(data) != w.NumberOfBases {
+		return fmt.Errorf("wrong number of bases, have %v, want %v", len(data), w.NumberOfBases)
+	}
+	if _, err := w.writer.Write(getbytes.FromInt32(int32(recordSamples))); err != nil {
```

Comment:
I'm not saying I know how to improve on this myself for sure, but this seems to have too many `Write` calls and error checks. Yes, yes, premature optimization. But can't we string together the various `[]byte` slices returned by `get bytes.FromXYZ` with `append` and make one `Write` and one error check?

Reply:
I'm using a buffered writer, I think it's equivalent to doing all that myself, except it works across calls to `WriteRecord`